js_modules/dagster-ui/packages/ui-core/src/asset-data/AssetLiveDataProvider.tsx

```diff
@@ -0,0 +1,292 @@
+    fragment: ASSET_LATEST_INFO_FRAGMENT,
+    fragmentName: 'AssetLatestInfoFragment',
+    id: `assetLatestInfoFragment-${uniqueId}`,
+  });
+  if (cachedAssetData && cachedLatestInfo) {
+    return {cachedAssetData, cachedLatestInfo};
+  } else {
+    return null;
+  }
+}
+
+async function _queryAssetKeys(client: ApolloClient<any>, assetKeys: AssetKeyInput[]) {
+  const {data} = await client.query<AssetGraphLiveQuery, AssetGraphLiveQueryVariables>({
+    query: ASSETS_GRAPH_LIVE_QUERY,
+    fetchPolicy: 'network-only',
+    variables: {
+      assetKeys,
+    },
+  });
+  data.assetNodes.forEach((assetNode) => {
+    _assetKeyToUniqueId[JSON.stringify(assetNode.assetKey.path)] = assetNode.id;
+    client.writeFragment({
+      fragment: ASSET_NODE_LIVE_FRAGMENT,
+      fragmentName: 'AssetNodeLiveFragment',
+      id: `assetNodeLiveFragment-${assetNode.id}`,
+      data: assetNode,
+    });
+  });
+  data.assetsLatestInfo.forEach((assetLatestInfo) => {
+    _assetKeyToUniqueId[JSON.stringify(assetLatestInfo.assetKey.path)] = assetLatestInfo.id;
+    client.writeFragment({
+      fragment: ASSET_LATEST_INFO_FRAGMENT,
+      fragmentName: 'AssetLatestInfoFragment',
+      id: `assetLatestInfoFragment-${assetLatestInfo.id}`,
+      data: assetLatestInfo,
+    });
+  });
+}
+
+// How many assets to fetch at once
+const BATCH_SIZE = 50;
+
+// Milliseconds we wait until sending a batched query
+const BATCHING_INTERVAL = 100;
+
+export const SUBSCRIPTION_IDLE_POLL_RATE = 30 * 1000;
+// const SUBSCRIPTION_MAX_POLL_RATE = 2 * 1000;
+
+type DataForNodeListener = (stringKey: string, data: LiveDataForNode) => void;
+
+const AssetLiveDataContext = React.createContext<{
+  useAssetNodeLiveData: (keys: AssetKeyInput[]) => Record<string, LiveDataForNode | null>;
+}>({
+  useAssetNodeLiveData: () => ({}),
+});
+
+// Map of asset keys to their last fetched time and last requested time
+const lastFetchedOrRequested: Record<
+  string,
+  {fetched: number; requested?: undefined} | {requested: number; fetched?: undefined} | null
+> = {};
+
+export const AssetLiveDataProvider = ({children}: {children: React.ReactNode}) => {
+  const [needsImmediateFetch, setNeedsImmediateFetch] = React.useState<boolean>(false);
+
+  const client = useApolloClient();
+
+  function useAssetNodeLiveData(assetKeys: AssetKeyInput[]) {
+    const [data, setData] = React.useState<Record<string, LiveDataForNode | null>>({});
+
+    React.useEffect(() => {
+      const setDataSingle = (stringKey: string, assetData: LiveDataForNode) => {
+        setData((data) => {
+          return {...data, [stringKey]: assetData};
+        });
+      };
+      assetKeys.forEach((key) => {
+        _subscribeToAssetKey(client, key, setDataSingle, setNeedsImmediateFetch);
+      });
+      return () => {
+        assetKeys.forEach((key) => {
+          _unsubscribeToAssetKey(key, setDataSingle);
+        });
+      };
+    }, [assetKeys]);
+
+    return data;
+  }
```

Comment:
It doesn't seem like this needs to be inside the provider - the context could provide a `setNeedsImmediateFetch`? 

It's kinda weird that this hook doesn't register the set of observed assets with the context - the context doesn't really have any context in it?